src/iosrv/drpc_ras.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * (C) Copyright 2021 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * This file is part of the DAOS server. It implements dRPC client RAS event
+ * functionality for communicating events to the control plane.
+ */
+
+#define D_LOGFAC DD_FAC(server)
+
+#include <daos/drpc_modules.h>
+#include <daos_srv/ras.h>
+#include <sys/utsname.h>
+#include "event.pb-c.h"
+#include "drpc_internal.h"
+#include "srv_internal.h"
+
+static void
+free_ras(Shared__RASEvent *evt)
+{
+	if (evt->obj_id)
+		D_FREE(evt->obj_id);
+	if (evt->pool_uuid)
+		D_FREE(evt->pool_uuid);
+	if (evt->cont_uuid)
+		D_FREE(evt->cont_uuid);
+	if (evt->hostname)
+		D_FREE(evt->hostname);
+	if (evt->timestamp)
+		D_FREE(evt->timestamp);
+}
+
+static int
+init_ras(ras_event_t id, char *msg, ras_type_t type, ras_sev_t sev, char *hid,
+	 d_rank_t *rank, char *jid, uuid_t *puuid, uuid_t *cuuid,
+	 daos_obj_id_t *oid, char *cop, Shared__RASEvent *evt)
+{
+	struct timeval	 tv;
+	struct tm	*tm;
+	int		 rc;
+
+	evt->timestamp = NULL;
+	evt->hostname = NULL;
+	evt->msg = NULL;
+	/* max value indicates nil rank */
+	evt->rank = UINT32_MAX;
+	evt->hw_id = NULL;
+	evt->proc_id = NULL;
+	evt->thread_id = NULL;
+	evt->job_id = NULL;
+	evt->pool_uuid = NULL;
+	evt->cont_uuid = NULL;
+	evt->obj_id = NULL;
+	evt->ctl_op = NULL;
```

Comment:
It would be more efficient to use "else" before for each if statement instead of initializing each individual fields to NULL.

Reply:
The reason I initialised like this is to avoid memleaks, if I make the change you suggested I get an invalid free warning because in free_ras evt->obj_id is uninitialised so D_FREE gets called. This looks maybe like a bug in the proto init auto generated function for RASEvent but I'm not sure.

Reply:
I'm restructuring to make it work